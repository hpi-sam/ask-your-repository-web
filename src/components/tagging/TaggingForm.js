// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { flashSuccessMessage } from 'redux-flash';
import TagSelector from './TagSelector';
import SaveButton from '../utility/SaveButton';
import type { Image } from '../../models/Image';
import type { AppState } from '../../state/AppState';
import fetchCreateTags from '../../requests/tagRequests';
import './TaggingForm.scss';

type Props = {
  image: Image,
  dispatch: Function,
};

type State = {
  isSubmitted: boolean,
};

class TaggingForm extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isSubmitted: false,
    };
  }

  handleSubmit = async () => {
    const { image, dispatch } = this.props;

    try {
      await fetchCreateTags(image.id, image.tags);
      dispatch(flashSuccessMessage('Your image was sucessfully uploaded.'));
      this.setState({ isSubmitted: true });
    } catch (error) {
      // TODO: Error Handling
    }
  };

  render() {
    if (this.state.isSubmitted) {
      return <Redirect to="/upload" />;
    }

    return (
      <div className="TaggingForm">
        <div className="TaggingForm__title">
          Tag your image!
        </div>
        <div className="TaggingForm__info">
          Type in a tag - Hit enter &#9166; - Repeat
        </div>
        <TagSelector />
        <SaveButton
          onClick={this.handleSubmit}
          className="TaggingForm__button"
        >
          Save
        </SaveButton>
      </div>
    );
  }
}

const mapStateToProps: Object = (state: AppState) => ({
  image: state.image,
});

export default connect(mapStateToProps)(TaggingForm);
