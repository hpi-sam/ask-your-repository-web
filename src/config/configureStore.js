// @flow
import { applyMiddleware, createStore } from 'redux';
import type { Dispatch } from 'redux';
import { persistStore, persistReducer } from 'redux-persist-erksch';
import storage from 'redux-persist-erksch/lib/storage';
import { routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { middleware as flashMiddleware } from 'redux-flash';
import thunk from 'redux-thunk';
import socketioMiddleware from '../middleware/socketioMiddleware';
import axiosAuthMiddleware from './axiosAuthMiddleware';
import createRootReducer from '../state/rootReducer';
import type { AppState } from '../state/AppState';
import type { Action } from '../state/Action';

export const history = createBrowserHistory();

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['activeTeam', 'presentationMode', 'auth'],
};

function configureStore() {
  let middleware = [
    thunk,
    flashMiddleware(),
    routerMiddleware(history),
    socketioMiddleware(),
    axiosAuthMiddleware(),
  ];

  if (process.env.NODE_ENV === 'development') {
    const { logger } = require('redux-logger');
    middleware = [...middleware, logger];
  }

  const rootReducer = createRootReducer(history);
  const persistedReducer: any = persistReducer(persistConfig, rootReducer);

  const store = createStore<AppState, Action, Dispatch<Action>>(
    persistedReducer,
    applyMiddleware(...middleware),
  );

  const persistor = persistStore(store);

  return { store, persistor };
}

export default configureStore;
