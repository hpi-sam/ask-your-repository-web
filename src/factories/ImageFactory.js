// @flow
import _ from 'lodash';
import faker from 'faker';
import type { Image } from '../models/Image';

class ImageFactory {
  static createDummyImage(tagAmount: number = 3): Image {
    return {
      id: faker.random.number(),
      url: faker.image.dataUri(),
      tags: _.times(tagAmount, () => faker.lorem.slug()),
    };
  }

  static createStaticDummyImage(): Image {
    return {
      id: 3826,
      url: 'data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%20%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%20%20%3Ctext%20x%3D%220%22%20y%3D%2220%22%20font-size%3D%2220%22%20text-anchor%3D%22start%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%20%3C%2Fsvg%3E',
      tags: ['nemo-dolorum-iusto', 'aut-quia-atque', 'et-velit-dolorem'],
    };
  }
}

export default ImageFactory;
