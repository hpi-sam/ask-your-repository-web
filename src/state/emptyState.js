// @flow
import type { AppState } from './AppState';

const emptyState: AppState = {
  flash: { messages: [] },
  image: null,
};

export default emptyState;
