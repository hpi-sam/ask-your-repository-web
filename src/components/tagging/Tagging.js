// @flow
import React from 'react';
import { connect } from 'react-redux';
import TaggingForm from './TaggingForm';
import TaggingImagePreview from './TaggingImagePreview';
import type { Image } from '../../models/Image';
import type { AppState } from '../../state/AppState';
import './Tagging.scss';

type Props = {
  image: Image,
};

function Tagging(props: Props) {
  if (!props.image) return null;

  return (
    <div className="Tagging">
      <div className="Tagging__inner">
        <TaggingForm />
        <TaggingImagePreview image={props.image} />
      </div>
    </div>
  );
}

const mapStateToProps = (state: AppState) => ({
  image: state.image,
});

export default connect(mapStateToProps)(Tagging);
