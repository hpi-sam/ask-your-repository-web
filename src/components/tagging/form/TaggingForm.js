// @flow
import React from 'react';
import type { Node } from 'react';
import TagSuggestions from './TagSuggestions';
import type { Tag } from '../../../models/Tag';
import './TaggingForm.scss';

type Props = {
  addTag: (tag: Tag) => void,
  isMultiTaggingEnabled: boolean,
  tagSelector: Node,
  tags: Array<Tag>,
};

function TaggingForm(props: Props) {
  const {
    tags,
    addTag,
    isMultiTaggingEnabled,
    tagSelector,
  } = props;

  return (
    <div className="TaggingForm">
      <div className="TaggingForm__title">
        Tag your image!
      </div>
      <div className="TaggingForm__info">
        Suggestions - Type number to add
      </div>
      <TagSuggestions
        tags={tags}
        addTag={addTag}
      />
      <div className="TaggingForm__info">
        Type in a tag - Hit enter &#9166; - Repeat
      </div>
      {tagSelector}
      {isMultiTaggingEnabled && (
        <div className="TaggingForm__tip">
          <div className="TaggingForm__tip__title">
            Multi-Tagging
          </div>
          You can press a tag to apply the tag
          <br />
          to all your uploaded images at once.
          <br />
          Multi tags will be highlighted with an extra border.
        </div>
      )}
    </div>
  );
}

TaggingForm.defaultProps = {
  enableMultiTagging: false,
  addMultiTag: () => {},
  removeMultiTag: () => {},
  multiTags: [],
};

export default TaggingForm;
