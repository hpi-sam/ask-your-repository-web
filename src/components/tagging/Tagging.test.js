// @flow
import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import Tagging from './Tagging';
import TaggingForm from './TaggingForm';
import TaggingImagePreview from './TaggingImagePreview';
import ImageFactory from '../../factories/ImageFactory';
import emptyState from '../../state/emptyState';

const mockStore = configureStore();

describe('<Tagging />', () => {
  let wrapper;
  let store;
  let state = emptyState;

  beforeEach(() => {
    store = mockStore(() => state);
    wrapper = shallow(<Tagging store={store} />).dive();
  });

  describe('no image is set', () => {
    it('should render nothing', () => {
      expect(wrapper.html()).toEqual(null);
    });
  });

  describe('image is set', () => {
    beforeAll(() => {
      state = { ...emptyState, image: ImageFactory.createStaticDummyImage() };
    });

    it('renders correctly', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders a TaggingForm component', () => {
      expect(wrapper.find(TaggingForm).exists()).toBeTruthy();
    });

    it('renders a TaggingImagePreview component', () => {
      expect(wrapper.find(TaggingImagePreview).exists()).toBeTruthy();
    });
  });
});
