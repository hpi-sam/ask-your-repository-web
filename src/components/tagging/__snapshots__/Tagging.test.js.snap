// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tagging /> image is set renders correctly 1`] = `
<div
  className="Tagging"
>
  <div
    className="Tagging__inner"
  >
    <Connect(TaggingForm) />
    <TaggingImagePreview
      image={
        Object {
          "id": 3826,
          "tags": Array [
            "nemo-dolorum-iusto",
            "aut-quia-atque",
            "et-velit-dolorem",
          ],
          "url": "data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%20%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%20%20%3Ctext%20x%3D%220%22%20y%3D%2220%22%20font-size%3D%2220%22%20text-anchor%3D%22start%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%20%3C%2Fsvg%3E",
        }
      }
    />
  </div>
</div>
`;
